### PR TITLE
comdb2_files: A system table to allow making a copy of data directory (WIP)

### PR DESCRIPTION
**Pull request by nirbhayc**
An async replicant can use this system table to inherit the initial state.

Changes have been made so that comdb2 will automatically try to copy the contents of data directory of the source database if it finds 'replicate_from' in its lrl.

Credit: @mponomar

Signed-off-by: Nirbhay Choubey <nchoubey@bloomberg.net>

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ db/phys_rep.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/df6c3b07/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/df6c3b07/smoketest.log'>Error &#x26A0; </a>.<br>Cbuild submission: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/df6c3b07/robomark.txt'>Error &#x26A0; </a>.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/df6c3b07/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/df6c3b07/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2726'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ db/phys_rep.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3bd0746d/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3bd0746d/smoketest.log'>Error &#x26A0; </a>.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3bd0746d/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3bd0746d/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2726'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ db/phys_rep.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/7066ad5a/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/7066ad5a/smoketest.log'>Error &#x26A0; </a>.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/7066ad5a/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/7066ad5a/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2726'>Resubmit</a></p>

**Comment by mponomar** [comment]
Quick thoughts - we rely on a couple of behaviors in comdb2ar (which are the reasons we don't use the system tar)

1. We need to pause log deletion while the copy is taking place because we may need lot entries to recover the copied files.
1. comdb2ar computes checksums to prevent torn page issues and retries on an checksum failures

The existing code will need to do something similar.  For (1) we can pause deletion for the duration of any queries that run against comdb2_files.  For (2) can can replicate comdb2ar's logic, (or move it into a library shared between comdb2 and comdb2ar).  A more correct option is to go through the db cache, but that'll be detrimental for performance.

File: db/comdb2.c [hunk 1]
```diff
@@ -3841,6 +3841,16 @@ static int init(int argc, char **argv)
        lrlname = create_default_lrl_file(dbname, gbl_dbdir);
     }
 
+    if (!gbl_create_mode && lrlname) {
+        int pre_process_physrep_options(const char *lrlname);
```
Comment (mponomar) [inline db/comdb2.c:3845]: Put this prototype in a shared header somewhere?

File: sqlite/ext/comdb2/files.c [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+static int filesFilter(
+  sqlite3_vtab_cursor *pVtabCursor, 
+  int idxNum, const char *idxStr,
+  int argc, sqlite3_value **argv
+){
+  systbl_files_cursor *pCur = (systbl_files_cursor *)pVtabCursor;
+  int i = 0;
+  int rc;
+
+  if( idxNum & 1 ){
```
Comment (mponomar) [inline sqlite/ext/comdb2/files.c:556]: Constants for these would be easier to remember.

**Pull request closed without merging**